### PR TITLE
Add the outward-bound part of the Cordova redirect flow to auth@exp

### PR DESCRIPTION
**Pull request by samhorlbeck**
This change handles the generation of the redirect URL including lots of platform-specific handling. This change allows `signInWithRedirect()` to successfully redirect to the oauth handler / IdP page. The event is not yet handled on return, however.

### CODE REVIEW

File: packages-exp/auth-exp/src/core/util/handler.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  redirectUrl?: string;
+  v: string;
+  providerId?: string;
+  scopes?: string;
+  customParameters?: string;
+  eventId?: string;
+  tid?: string;
+} & { [key: string]: string | undefined };
+
+export function _getRedirectUrl(
```
Comment (samhorlbeck) [inline packages-exp/auth-exp/src/core/util/handler.ts:56]: This is moved from `platform_browser/popup_redirect`. The only change is that it now has an "additionalParams" parameter

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  return chars.join('');
+}
+
+/**
+ * Computes the SHA-256 of a session ID. The SubtleCrypto interface is only
+ * available in "secure" contexts, which covers Cordova (which is served on a file
+ * protocol).
+ */
+async function computeSha256(sessionId: string): Promise<string> {
+  const encoder = new TextEncoder();
```
Comment (yuchenshi) [inline packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:185]: What do we do for IE11? https://caniuse.com/textencoder says no support

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  const buf = await crypto.subtle.digest('SHA-256', encoder.encode(sessionId));
```
Comment (yuchenshi) [inline packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:186]: In IE11, this returns `CryptoOperation` (not Promise) and `await`-ing this won't give the right result.

See: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  const arr = new Uint8Array(buf);
+
+  let hexString = '';
```
Comment (yuchenshi) [inline packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:189]: nit: MDN has a more concise implementation here: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#converting_a_digest_to_a_hex_string

It involves `Array.from` but I think performance shouldn't be a concern here -- code is not on a critical path and Array size is constant.

**Review by yuchenshi — APPROVED** [review]

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts (hunk 3 above)
Comment (yuchenshi) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:186]: FYI: `CryptoOperation` is from an earlier draft here: https://www.w3.org/TR/2013/WD-WebCryptoAPI-20130108/#cryptooperation-interface

tl;dr: we need to attach `oncomplete` / `onerror` handlers instead of `await`.

File: packages-exp/auth-exp/karma.conf.js [hunk 5]
```diff
@@ -35,6 +35,8 @@ function getTestFiles(argv) {
     return ['src/**/*.test.ts', 'test/helpers/**/*.test.ts'];
   } else if (argv.integration) {
     return ['test/integration/**/*.test.ts'];
+  } else if (argv.cordova) {
+    return ['src/platform_cordova/**/*.test.ts'];
```
[the remark this answers is not in the document]
Comment (samhorlbeck) [reply packages-exp/auth-exp/karma.conf.js:39]: Which other tests?

File: packages-exp/auth-exp/src/model/popup_redirect.ts [hunk 6]
```diff
@@ -86,7 +86,7 @@ export interface PopupRedirectResolver extends externs.PopupRedirectResolver {
     provider: externs.AuthProvider,
     authType: AuthEventType,
     eventId?: string
-  ): Promise<never>;
+  ): Promise<unknown>;
```
[the remark this answers is not in the document]
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/model/popup_redirect.ts:89]: In Cordova environments, the redirect does not necessarily end the activity, so a call to `signInWithRedirect()` can actually resolve

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.test.ts [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+  _checkCordovaConfiguration,
+  _generateHandlerUrl,
+  _generateNewEvent,
+  _performRedirect
+} from './utils';
+import { AuthEvent, AuthEventType } from '../../model/popup_redirect';
+import { GoogleAuthProvider } from '../../core/providers/google';
+import { AuthProvider } from '../../../internal';
+
+enum UA {
```
[the remark this answers is not in the document]
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.test.ts:34]: I had that originally but I liked this better lol

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      (BuildInfo as Record<string, string>).displayName = 'This is my app';
+      const params = getParams(
+        await _generateHandlerUrl(auth, event, provider)
+      );
+      expect(params.get('appDisplayName')).to.eq('This is my app');
+    });
+  });
+
+  describe('_performRedirect', () => {
+    let isBrowsertabAvailable: boolean;
```
[the remark this answers is not in the document]
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.test.ts:205]: Only one of these has it set to true, the rest to false

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:185]: I didn't realize Cordova supported IE11 :(

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts (hunk 3 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts:186]: Ya I didn't realize Cordova supported IE11. I'm going to add this as a TODO but since I can't actually test this at the moment anyway I don't want to block on IE11 support here

File: packages-exp/auth-exp/src/core/util/handler.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    provider.setDefaultLanguage(auth.languageCode);
+    params.providerId = provider.providerId || '';
+    if (!isEmpty(provider.getCustomParameters())) {
+      params.customParameters = JSON.stringify(provider.getCustomParameters());
+    }
+    const scopes = provider.getScopes().filter(scope => scope !== '');
+    if (scopes.length > 0) {
+      params.scopes = scopes.join(',');
+    }
+    // TODO set additionalParams?
```
[the remark this answers is not in the document]
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/core/util/handler.ts]: Keeping this around to refer to the provider

File: packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts (hunk 2 above)
Comment (samhorlbeck) [reply packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts]: Added a function to fallback to a simple encoding function
